### ISSUE TITLE
C++ cv::VideoCapture.open(0) always return false on Android

### ISSUE THREAD

**Issue by omatrot** [opencv/opencv#11952]
##### System information (version)

- OpenCV => 3.4.1
- Operating System macOS High Sierra 10.13.6 / Platform => Android 7.1.1 API 25
- Compiler => Android NDK

###### Context

I'm writing a cross platform OpenCV based C++ library. The consuming code is a React Native Application through a [react native native module](https://facebook.github.io/react-native/docs/native-modules-android.html).

To be perfectly clear, there is no access from Java Code to C++ OpenCV on Android. There are events with the result of the OpenCV C++ code sent to Javascript through the react native bridge.

My native library is compiled on Android as a SHARED library. It is dynamically linked to the libopencv_world.so that is produced by the compilation of OpenCV C++ for Android.

###### What it does

Basically, it opens the device's default camera and take snapshots.

###### The outcome

This code is then ran on iOS and Android.

This is working perfectly well on iOS.
It fails on Android.

Here is the failing part of C++ code on Adndroid:

##### Steps to reproduce

```.cpp
// cap is a cv::VideoCapture object    
if (cap.open(0))
        {
            cap.set(cv::CAP_PROP_FRAME_WIDTH, CAM_WIDTH);
            cap.set(cv::CAP_PROP_FRAME_HEIGHT, CAM_HEIGHT);
        }
        else
        {
            reject("false", "cap.open(0) returned false");
        }
```

**Comment by alalek** [comment]
C++ `VideoCapture` camera interface is not supported on Android platform.

**Comment by omatrot** [comment]
So basically, what are my options on this platform ?
FYI, [I asked](http://answers.opencv.org/question/195522/c-ndk-cvvideocaptureopen0-return-false-on-android/) on the Q/A forum. No answer.

**Comment by mshabunin** [comment]
Native camera interface has been added in Android 7: https://developer.android.com/ndk/guides/stable_apis

**Comment by potrykus** [comment]
Following the [previous](https://github.com/opencv/opencv/issues/11952#issuecomment-404541724) comment, I think it is appropriate to turn this into a [important] feature request.

[Here](https://developer.android.com/ndk/guides/stable_apis#a24) is a relevant blurb on the Android 7 [API](https://developer.android.com/ndk/reference/group/camera) (level 24):

> The native camera API provides the native equivalent of the android.hardware.camera2 classes in the Java programming language. The native camera API lets you perform fine-grained photo capture and processing in your native code.

Instability, or non-constancy cross-device, was an issue for OpenCV implementations from 2.4 on.  This is a stable API, so there should be no impediment for OpenCV implementation.

Given android market shares, having (at least) the same capability as on iOS is hugely valuable.

This could be an experimental development, first just getting the [CAP_ANDROID](https://docs.opencv.org/3.4.2/d4/d15/group__videoio__flags__base.html#gga023786be1ee68a9105bf2e48c700294da0a3b80ec023425a27431991500d26847) flag usable at the level of the videoCapture [tutorial](https://docs.opencv.org/3.4.2/d5/dc4/tutorial_video_input_psnr_ssim.html) (e.g., through the >> operator).  Further exposure of native-camera control bits can occur in later (minor) versions.  In some sense, the development would never have to leave "experimental" status (for the foreseeable future).

I don't know iOS native AVFoundation, nor do I know the openCV iOS videoCapture implementation, but given the Swift AVFoundation's handling of callbacks, I imagine a port of the openCV iOS videoCapture implementation is more than "just feasible."  Further, the Android native-camera (level 24) API is actually cleaner than that of iOS.  This makes implementation even easier.  [Here](https://github.com/justinjoy/native-camera2) is a nice example of the 4 (only) callbacks necessary for hooking into the native-camera video stream: [https://github.com/justinjoy/native-camera2/blob/master/app/src/main/jni/native-camera2-jni.cpp](https://github.com/justinjoy/native-camera2/blob/master/app/src/main/jni/native-camera2-jni.cpp)  N.B., this code should not be intimidating (esp. for C++!): it is almost 100% glue code.  

(Note SDK N = Nougat = Android 7 = API 24.)

As a last point, justinjoy's native-camera2 project here has the point of gluing the native camera into a java [driver](https://github.com/justinjoy/native-camera2/blob/master/app/src/main/java/org/freedesktop/nativecamera2/NativeCamera2.java) through JNI.  The parallel with the OpenCV goal is pretty strong: replace "through-JNI driver calls" with "through OpenCV impl to public cv::videoCapture API".

**Comment by lapinozz** [comment]
Would this let us use cv::VideoCapture on Android with both Camera2 and Camera1(LEGACY) cameras? That'd be awesome!

**Comment by githubdoramon** [comment]
I am a little lost here... does that means that for now we can't use OpenCV to capture camera images on Android newest versions?

**Comment by email2jie** [comment]
Would like to know if there are any other options at the moment as well.

**Comment by mshabunin** [comment]
You can use [camera2](https://developer.android.com/reference/android/hardware/camera2/package-summary) in Java (since 5.0) or [NDK Camera](https://developer.android.com/ndk/reference/group/camera) in C++ (since 7.0).

**Comment by komakai** [comment]
#14005 implements a native back-end for video file decoding (but not camera capture)
It requires API LEVEL to be set to 21 for example by building the OpenCV Android SDK with 
`platforms/android/build_sdk.py --config ndk-18-api-level-21.config.py {...other arguments}`

A possible improvement would be dynamically checking the presence of libmediandk.so at runtime using `dlopen` and then use `dlsym` to lookup all the necessary APIs. This approach would mean you could still get the native file decoding without having to set the API LEVEL to 21. You could also then add a Java implementation for pre-Lollipop Androids that would kick in if the native implementation wasn't available.
A Java implementation should be pretty simple using the following for guidance:
https://github.com/googlesamples/android-BasicMediaDecoder
https://github.com/google/grafika (I haven't checked but OffscreenSurface.java might be useful)

**Comment by alalek** [comment]
@komakai Thank you for adding support for reading of video files!

> dynamically checking the presence of libmediandk.so

There is attempt to reduce required external dependencies and make them optional with plugins approach (see #13677).

**Comment by zmdsjtu** [comment]
Is it possible to open camera using cv::videoCapture(0) in android now?

**Comment by komakai** [comment]
> Is it possible to open camera using cv::videoCapture(0) in android now?

Not yet.
If you are interested in this feature then please try out the branch below and let me know if it works for you:
https://github.com/komakai/opencv/tree/android-ndk-camera

**Comment by avaish1** [comment]
I am planning to use  this link to make videocapture in android application using opencv. will update in coming days.

**Comment by LankyBin-lb** [comment]
Is there any ways to get camera buffer to instead "capture >> mat",i'v tried android camera buffer.But the data of buffer is ataxic when I use "Mat frame(720, 1280, CV_8UC1, buffer)"

**Comment by AI-Decay** [comment]
Did this feature appear after version version: 4.1.1 or higher?

**Comment by komakai** [comment]
@AI-Decay if you are interested in this feature then please try out https://github.com/komakai/opencv/tree/android-ndk-camera
and give feedback.

**Comment by jumpy88** [comment]
Has anyone tested @komakai code?

**Comment by jumpy88** [comment]
Hello @komakai, I'm interested in camera capture feature and I'd like to try your branch. I'm working in linux. I've android-ndk-r21d installed but I can't figure out how to build your source code. Can you give me some hints?
Thank you in advance.

**Comment by komakai** [comment]
Hello @jumpy88
The branch was quite old so I have synced it with master: 9315e45
In order to build of course you need the Android SDK and Android NDK installed and you need the ANDROID_SDK environment variable set up to point to the Android SDK directory.
And of course you need to check out my branch, for example like:
`git clone https://github.com/komakai/opencv.git`
`git co android-ndk-camera`
Then the following should build the OpenCV Android with NDK Camera support (call from the directory below the opencv checkout)
`./opencv/platforms/android/build_sdk.py --use_android_buildtools --no_samples_build --config ./opencv/platforms/android/ndk-18-api-level-24.config.py android_build opencv`

I will update later with info how to actually test once you get a build out

**Comment by jumpy88** [comment]
Thank you @komakai for your support.
I just need to build libopencv_java4.so to link it in a Godot application, so I'll not need to write any java code line. Do I also need Android SDK and not only Android NDK?
I see in python script name "ndk-18". Does it mean ndk r21 is still not supported and should I install a previous version?

**Comment by komakai** [comment]
Hello @jumpy88 

> Do I also need Android SDK and not only Android NDK?

Currently `platforms/android/build_sdk.py`  has no option to just build the native part so you will need the Android SDK as well.

> I see in python script name "ndk-18". Does it mean ndk r21 is still not supported and should I install a previous version?

It's no problem to use a newer NDK - the number in the script name was originally required because earlier versions of the NDK had different toolchain names. The "ndk-18" script doesn't specify a toolchain name and the build script just defaults to `clang`. As a result of this the "ndk-18" script works with newer NDKs as well.

For my own testing I created an app here: https://github.com/komakai/NdkCamtest feel free to use that to test as well.
In my testing I found 2 issues:
 - the feed has red and blue channels flipped (this might be the expected behaviour - would need to check the spec)
 - stopping and restarting doesn't work - looks like some extra clean-up is required when stopping the camera

**Comment by jumpy88** [comment]
Hello @komakai, I'm close but something still goes wrong: it builds up to 880/890 (or to the end but I can't see because is too fast) but then it fails with this error

>FAILURE: Build failed with an exception.

\* What went wrong:
A problem occurred configuring project ':face-detection'.
\> NDK not configured. 
Download it with SDK manager.

Obviously I installed NDK using SDK Manager inside Android Studio and I tried to set it both as ANDROID_NDK environment variable and --ndk_path parameter but nothing seems to work. What should I do to make it work?

**Comment by komakai** [comment]
I think if you pass ` --no_samples_build` then that error will go away
Alternatively try setting `ANDROID_NDK_HOME` - I think that also should solve it

**Comment by jumpy88** [comment]
`--no_samples_build` is not accepted. It makes the script print the usage. I'm trying different configurations for ANDROID_NDK, ANDROID_NDK_ROOT, ANDROID_NDK_HOME, --sdk_path, --ndk_path...but can't find a working setup. Where are they supposed to point to?

**Comment by komakai** [comment]
> --no_samples_build is not accepted.

Can you do `git rev-parse HEAD` and check the output starts 9315e45  ?

**Comment by jumpy88** [comment]
Ops...you are right, I messed up my local repository after your recent merge with master. I have redownloaded the repository and checked out to the right branch (and checked the commit hash is correct 9315e45b54d4cb4fd14f021a29732946460265d9). Now it accepts --no_samples_build but I'm still doing something wrong:
>FAILURE: Build failed with an exception.
>
>\* What went wrong:
>A problem occurred configuring project ':opencv'.
>\> java.lang.NullPointerException (no error message)

**Comment by komakai** [comment]
Hmm - I believe that error happens when gradle can't find your JDK. Make sure you have the JDK installed and `JAVA_HOME` set

**Comment by jumpy88** [comment]
Ok, I solved this by adding --sdk_path and --ndk_path together with --no_samples_build. Now I've my four libopencv_java4.so files (one for each ABI). I'll make you know if they do the work for my project.
Thank you for your help

**Comment by komakai** [comment]
@jumpy88 OK let me know how you get on - please send LogCat output if you hit any issues

**Comment by jumpy88** [comment]
@komakai my apk starts but it doesn't show any data from USB camera (the same project linked to x86_64 OpenCV library and exported for linux works fine). I also tried to require all permissions but, when I start it, it neither ask me for permission grant. I looked for a way to filter logcat output only to show my apk relevant information, but I couldn't find any solution. Do you have any suggestion to do it?

**Comment by komakai** [comment]
> any data from USB camera

Ah - a USB camera. Then I don't think this will work. Android NDK Camera API (which is what I am using to implement `VideoCapture.open()`) doesn't know about external USB cameras. I recommend you take a look at https://github.com/saki4510t/UVCCamera

**Comment by jumpy88** [comment]
Great, a weekend thrown in the rubbish (and uselessly stolen a lot of your time) 😆
Anyway, I learnt at least something new about OpenCV and Android C++ libraries building.
For completeness: color blob detect sample from OpenCV ask for camera permission but it says "It seems that your device does not support camera (or it is locked). Application will be closed."
>I recommend you take a look at https://github.com/saki4510t/UVCCamera

Yes, thank you for the suggestion. I already knew about it (it is the library used by the application linked above) and it was already my second planned attempt. It wasn't the first one because I'll also need OpenCV for computation, I know a little of OpenCV whereas never used UVCCamera and it seems not to be an active repository since October 2018.
Thank you so much @komakai for all of your support.

**Comment by jumpy88** [comment]
@komakai just to make you know: you could be interested in [Android Camera2 API](https://source.android.com/devices/camera/external-usb-cameras) for further development of OpenCV VideoCapture Android support. It seems to be able to read data from USB cameras.

**Comment by komakai** [comment]
@jumpy88 Although UVCCamera has not been active recently I was able to get it to work. Please check out:
https://github.com/komakai/UVCCamera/tree/komakai-fixes

Feel free to create an issue over in that repository if you need help with UVCCamera

**Comment by jumpy88** [comment]
Oh, thanks @komakai, I was looking for an up to date UVCCamera fork.
See you there.

**Comment by jumpy88** [comment]
@komakai I couldn't create an issue in your repository, so I created it in the original one you forked. You find it here https://github.com/saki4510t/UVCCamera/issues/588

**Comment by nwrkbiz** [comment]
Hmm for me it works fine out of the box. At least if I use a cheap USB webcam which seems to be v4l2 compatible.

I cross compiled a completely statically linked executable using aarch64-linux-musl-g++ so no android specific code.

**Comment by jumpy88** [comment]
@nwrkbiz do you mean plain C++ cross compiled source (the linked simple-webcam project)? How did run it without an apk?

**Comment by nwrkbiz** [comment]
> @nwrkbiz do you mean plain C++ cross compiled source (the linked simple-webcam project)? How did run it without an apk?

Yes.

I did wrap the executable in following apk:
https://github.com/nwrkbiz/android-xserver

**Comment by nwrkbiz** [comment]
If you have a rooted smartphone you can also perform these steps manually:

* Install any terminal emulator (i.e. connectbot).
* Install any android xserver (i.e. https://github.com/nwrkbiz/android-xserver).
* Download the statically linked simple-webcam executable matching your cpu architecture fromhttps://github.com/nwrkbiz/simple-webcam (most probably aarch64 or armhf)
* Start your android xserver and keep it running in the background
* Start a local terminal root session
* Copy the simple-webcam executable to /data using cp
* chmod u+x /data/simple-webcam
* export DISPLAY=127.0.0.1:0
* finally execute: /data/simple-webcam --device <<cam_id>>
* Switch back to android-xserver

**Comment by jumpy88** [comment]
One more question: did you link against libopencv_java4.so?

**Comment by nwrkbiz** [comment]
> One more question: did you link against libopencv_java4.so?

No, I linked against all the opencv libraries you can see within this makefile: https://github.com/nwrkbiz/simple-webcam/blob/main/Makefile

And these libraries were built with the help of: https://github.com/nwrkbiz/static-build

**Comment by jumpy88** [comment]
Ok, so I don't think this fits my case: my device is not rooted and the usb device is not mounted as /dev/videox. UVCCamera accede to it through /dev/bus/00x/00y. Thank you anyway for sharing these information, it's interesting if OpenCV started supporting VideoCapture for Android. Maybe @alalek could confirm this.

**Comment by nwrkbiz** [comment]
> Ok, so I don't think this fits my case: my device is not rooted and the usb device is not mounted as /dev/videox. UVCCamera accede to it through /dev/bus/00x/00y. Thank you anyway for sharing these information, it's interesting if OpenCV started supporting VideoCapture for Android. Maybe @alalek could confirm this.

Having no root should not be an issue. Hmm,... what Android version do you have? Do you have any /dev/videoX devices? Did you try different USB-Webcams?

**Comment by jumpy88** [comment]
`$ getprop ro.build.version.release`
>7.1.1

`$ getprop ro.build.version.sdk`
>25

I can't `ls` them but I can see /dev/video0 - 6 exist (they return `permission denied`, /dev/video7, 8...return `no such file or directory`). I tried with 2 different USB cameras, but no one of them seems to add a /dev/videoX, despite I can accede them with [this application](https://play.google.com/store/apps/details?id=com.shenyaocn.android.usbcamera) (they only seem to appear as /dev/bus/usb/001/00X devices)

ps what does it mean "akk"?

**Comment by nwrkbiz** [comment]
> ps what does it mean "akk"?

Sorry typo, I meant *all

**Comment by jumpy88** [comment]
Don't worry, I just thought it was some strange slang and I was missing something 😂😆
Anyway, the device is an Oculus Quest, which has 4 physical built-in cameras. I don't know why there are 6 /dev/videoX. Also, I don't know what happens with other not rooted Android devices, but maybe OQ doesn't add /dev/videoX for USB cameras to avoid messing up the localization.

**Comment by nwrkbiz** [comment]
@jumpy88 I am very sorry if I caused conusion, but saw that I granted my wrapper app root permissions, thats why it worked out for me.

**Comment by jumpy88** [comment]
@nwrkbiz no problem, we are investigating, any clue can be useful 🙂

**Comment by splinterwhatever** [comment]
Hey there!
@komakai the combination of your 9315e45 commit as well as the python build script is a thing of beauty!  Massive thanks for that.  I recently hit the lack of access to the camera issue and happened across this thread - your solution which works a treat!  Thanks also to @jumpy88 for documenting your journey, it was informative and helpful too.

I can confirm that when backgrounding my test app and returning I am unable to revive my camera stream - it appears frozen.  Is this what you meant by this @komakai?:
   o "stopping and restarting doesn't work - looks like some extra clean-up is required when stopping the camera"

**Comment by komakai** [comment]
@splinterwhatever thanks for testing and for your feedback. I'll have a go at fixing the stop/restart problem and then put together a PR.
BTW what is your use case? Is there some reason you prefer native camera access over the JavaCamera* classes?

**Comment by splinterwhatever** [comment]
Hey @komakai That would be awesome! Is there a way to help make this part of the official openCV repo?

To your question: I'm learning/working on a project relies upon some cross-platform frameworks, including openCV.  As such my language of choice for this project is C++, hence my aversion to Java thus far.

Beyond the stop/restart issue, the other thing I'm challenged with is capture resolution.  Try as I might, on Android I only seem to be able to pull in the max res for the camera - which understandably slows things down.  I'd like to be able to query the available resolutions per camera, then set it accordingly.  cv::VideoCapture.set and get seem to allow for this, but on my oneplus5 handset I've had no luck so far.

https://answers.opencv.org/question/186435/unable-to-set-video-capture-size/

**Comment by jumpy88** [comment]
Hello everyone, I'm trying to accede USB cameras in non-rooted Android devices in C++. I'm trying to use https://github.com/komakai/UVCCamera/tree/komakai-fixes but I can't find ho to set it up without using the Java wrapper in the examples provided with the library. If someone is interested or knows useful information we can discuss about it in a dedicated issue I've created
https://github.com/saki4510t/UVCCamera/issues/593
Thank you in advance for the contribution.

**Comment by komakai** [comment]
@splinterwhatever I have pushed a new commit with a fix for the hang when attempting to restart Camera capture.

> I'd like to be able to query the available resolutions per camera, then set it accordingly. cv::VideoCapture.set and get seem to allow for this, but on my oneplus5 handset I've had no luck so far.

Setting the resolution won't work yet because I haven't implemented the `AndroidCameraCapture::setProperty` function yet

**Comment by splinterwhatever** [comment]
@komakai Thats awesome. I was looking forward to checking it out - just got around to it.

Hmmm I must be doing something else wrong as I'm getting the same behaviour. Namely:
1. Open a test app and see the capture going nicely.
2. Return to home screen and then to another camera app (the standard android camera app in this case)
3. Navigate back to the previously running test app.
4. Note that there is no updating data from the cam device.

This is for:
`cv::VideoCapture cap {1}; //The self camera in my case
cv::Mat frame;`
doing a `cap>>frame`
displaying `frame`.

On a fresh run of the test app the camera behaves really well.
When following the steps, by the time you return to the test app apparently frame is empty with 0 columns and 0 rows.

Obviously not requesting/expecting anyone to debug my silliness. Just wanted to clarify the steps in case it hints at an issue with the restarting camera issue.


Also thanks for clarity on the resolution/AndroidCameraCapture::setProperty situation.  Would be superb if that got fixed too.

Let me know what I can do to help on either issue.

**Comment by komakai** [comment]
@jumpy88 
>  but I can't find ho to set it up without using the Java wrapper in the examples provided with the library. If someone is interested or knows useful information we can discuss about it in a dedicated issue I've created

Looking at the top answer here: https://stackoverflow.com/questions/22197425/low-level-usb-api-on-android-at-ndk-level-to-use-on-qt-android it looks like what you're trying to do isn't possible without using some Java, specifically you need to use the `UsbManager` to discover your devices

**Comment by jumpy88** [comment]
@komakai thanks for your reply! It's not what I was hoping but it's useful to know. I've also found some additional information https://forum.qt.io/topic/85176/crazy-frameworks-integration-and-i-m-stuck
I'll make you know if I find a way to use `UsbManager` inside C++ source code (through NDK maybe).

**Comment by wangrentao** [comment]
Is it possible to open camera using cv::videoCapture(0) in android now?

**Comment by Oxymore-coder** [comment]
> Is it possible to open camera using cv::videoCapture(0) in android now?

I second that question
